File: samples/cpp/automatic_speech_recognition/asr_streaming.cpp
// Copyright (C) 2026 Intel Corporation
// SPDX-License-Identifier: Apache-2.0

// Streaming ASR from a WAV file using ASRStreamingSession.
// Pushes the file in fixed-size segments without injecting artificial delays.
//
// Usage:
//   asr_streaming <MODEL_DIR> <WAV_FILE> [DEVICE] [CHUNK_SEC] [STEP_MS]
//
//   MODEL_DIR   — path to an OpenVINO Qwen3-ASR model directory
//   WAV_FILE    — mono 16 kHz PCM WAV file to transcribe
//   DEVICE      — OpenVINO device string (default: CPU)
//   CHUNK_SEC   — decode interval in seconds (default: 2.0)
//   STEP_MS     — chunk push interval in ms (default: 500)
//   --window-chunks N          — sliding window size in chunks, 0 = unbounded (default: 6)
//   --window-rollback-chunks N — unfixed rollback chunks within the window (default: 2)
//   --unbounded-prefix         — experimental, Qwen3-ASR only: keep audio bounded (per
//                                --window-chunks) but stop evicting the text prefix.
//   --repetition-penalty F     — generation repetition penalty, 1.0 = disabled (default: 1.0)
//   --no-repeat-ngram-size N   — forbid repeating any N-gram, 0 = disabled (default: 0)

#include <chrono>
#include <iomanip>
#include <iostream>

#ifdef _WIN32
#    define WIN32_LEAN_AND_MEAN
#    define NOMINMAX
#    include <windows.h>
#endif

#include "audio_utils.hpp"
#include "openvino/genai/automatic_speech_recognition/pipeline.hpp"

int main(int argc, char* argv[]) try {
#ifdef _WIN32
    // The model emits UTF-8 text; without this the Windows console renders non-ASCII bytes
    // (e.g. em-dashes) as mojibake under its default OEM codepage.
    SetConsoleOutputCP(CP_UTF8);
#endif

    if (argc < 3) {
        throw std::runtime_error(std::string{"Usage: "} + argv[0] +
                                 " <MODEL_DIR> <WAV_FILE> [DEVICE] [CHUNK_SEC] [STEP_MS] "
                                 "[--device DEVICE] [--chunk-sec SEC] [--step-ms MS] "
                                 "[--window-chunks N] [--window-rollback-chunks N] "
                                 "[--unbounded-prefix] "
                                 "[--repetition-penalty F] [--no-repeat-ngram-size N]");
    }

    const std::filesystem::path models_path = argv[1];
    const std::string wav_file = argv[2];
    std::string device = "CPU";
    float chunk_sec = 2.0f;
    int step_ms = 500;
    size_t window_chunks = 6;
    size_t window_rollback_chunks = 2;
    bool unbounded_prefix = false;
    float repetition_penalty = 1.0f;
    size_t no_repeat_ngram_size = 0;

    int positional_index = 0;
    for (int i = 3; i < argc; ++i) {
        const std::string arg = argv[i];

        if (arg == "--device") {
            if (i + 1 >= argc) {
                throw std::runtime_error("Missing value for --device");
            }
            device = argv[++i];
            continue;
        }

        if (arg == "--chunk-sec") {
            if (i + 1 >= argc) {
                throw std::runtime_error("Missing value for --chunk-sec");
            }
            chunk_sec = std::stof(argv[++i]);
            continue;
        }

        if (arg == "--step-ms") {
            if (i + 1 >= argc) {
                throw std::runtime_error("Missing value for --step-ms");
            }
            step_ms = std::stoi(argv[++i]);
            continue;
        }

        if (arg == "--window-chunks") {
            if (i + 1 >= argc) {
                throw std::runtime_error("Missing value for --window-chunks");
            }
            window_chunks = static_cast<size_t>(std::stoul(argv[++i]));
            continue;
        }

        if (arg == "--window-rollback-chunks") {
            if (i + 1 >= argc) {
                throw std::runtime_error("Missing value for --window-rollback-chunks");
            }
            window_rollback_chunks = static_cast<size_t>(std::stoul(argv[++i]));
            continue;
        }

        if (arg == "--unbounded-prefix") {
            unbounded_prefix = true;
            continue;
        }

        if (arg == "--repetition-penalty") {
            if (i + 1 >= argc) {
                throw std::runtime_error("Missing value for --repetition-penalty");
            }
            repetition_penalty = std::stof(argv[++i]);
            continue;
        }

        if (arg == "--no-repeat-ngram-size") {
            if (i + 1 >= argc) {
                throw std::runtime_error("Missing value for --no-repeat-ngram-size");
            }
            no_repeat_ngram_size = static_cast<size_t>(std::stoul(argv[++i]));
            continue;
        }

        if (positional_index >= 3) {
            throw std::runtime_error("Unexpected argument: " + arg);
        }

        switch (positional_index++) {
            case 0:
                device = arg;
                break;
            case 1:
                chunk_sec = std::stof(arg);
                break;
            case 2:
                step_ms = std::stoi(arg);
                break;
        }
    }

    std::cout << "Loading model from: " << models_path << " on " << device << "\n";
    ov::genai::ASRPipeline pipeline(models_path, device);

    ov::genai::ASRGenerationConfig gen_config = pipeline.get_generation_config();
    gen_config.max_new_tokens = 32;  // keep chunk decodes fast
    gen_config.repetition_penalty = repetition_penalty;
    if (no_repeat_ngram_size > 0) {
        gen_config.no_repeat_ngram_size = no_repeat_ngram_size;
    }

    ov::genai::ASRStreamingConfig streaming_config;
    streaming_config.chunk_size_sec = chunk_sec;
    streaming_config.warmup_chunks = 2;
    streaming_config.context_rollback_tokens = 5;
    streaming_config.window_chunk_num = window_chunks;
    streaming_config.window_rollback_chunk_num = window_rollback_chunks;
    streaming_config.unbounded_prefix = unbounded_prefix;

    std::cout << "Loading audio: " << wav_file << "\n";
    const ov::genai::RawSpeechInput wav = utils::audio::read_wav(wav_file);
    const size_t total_samples = wav.size();
    const float total_sec = static_cast<float>(total_samples) / 16000.0f;
    std::cout << std::fixed << std::setprecision(2)
              << "  Duration: " << total_sec << " s  (" << total_samples << " samples @ 16 kHz)\n\n";

    auto session = ov::genai::ASRStreamingSession(pipeline, streaming_config, gen_config);

    const size_t step_samples = static_cast<size_t>(step_ms * 16000 / 1000);
    const auto wall_start = std::chrono::steady_clock::now();

    int chunk_num = 0;
    for (size_t pos = 0; pos < total_samples; pos += step_samples) {
        const size_t end = std::min(pos + step_samples, total_samples);
        const std::vector<float> segment(wav.begin() + pos, wav.begin() + end);
        if (const auto result = session.push_chunk(segment)) {
            std::cout << chunk_num << " [partial] (" << result->language << ") +" << result->new_committed_text
                      << " [" << result->partial_text << "]\n";

            chunk_num++;
        }
    }

    const auto final_result = session.finish();
    if (!final_result.new_committed_text.empty()) {
        std::cout << chunk_num << "[partial] (" << final_result.language << ") +" << final_result.new_committed_text
                  << " []\n";
    }

    const auto wall_sec =
        std::chrono::duration<float>(std::chrono::steady_clock::now() - wall_start).count();
    std::cout << "\n[final ] (" << final_result.language << ") " << final_result.committed_text << "\n";
    std::cout << std::fixed << std::setprecision(2)
              << "\nTotal wall time: " << wall_sec << " s  (RTF " << wall_sec / total_sec << "x)\n";
    return 0;

} catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << "\n";
    return 1;
}

File: samples/python/automatic_speech_recognition/asr_streaming.py
#!/usr/bin/env python3
# Copyright (C) 2026 Intel Corporation
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

import argparse
import time
from typing import Any

import librosa
import numpy as np

try:
    import openvino_genai
except ModuleNotFoundError as exc:
    raise RuntimeError(
        "OpenVINO GenAI Python package is not installed in the active environment. "
        "Build or install the repo in this interpreter before running the streaming ASR sample."
    ) from exc

if not hasattr(openvino_genai, "ASRStreamingConfig"):
    raise RuntimeError(
        "The active openvino_genai package does not include the streaming ASR API. "
        "Rebuild/install the repo from this checkout into the current Python environment."
    )


SAMPLE_RATE = 16000


def read_wav(path: str) -> np.ndarray:
    audio, sr = librosa.load(path, sr=SAMPLE_RATE, mono=True)
    return np.asarray(audio, dtype=np.float32)


def print_result(result: Any, inline: bool) -> None:
    if result is None:
        return
    if inline:
        print(result.new_committed_text, end="", flush=True)
    else:
        print(f"[partial] ({result.language}) +{result.new_committed_text} [{result.partial_text}]")


def run_wav_streaming(pipe: Any, wav_path: str, cfg: Any,
                     inline: bool, block_ms: int = 250) -> Any:
    audio = read_wav(wav_path)
    session = pipe.create_streaming_session(cfg)

    block_size = max(1, int((block_ms / 1000.0) * SAMPLE_RATE))
    for start in range(0, len(audio), block_size):
        end = min(start + block_size, len(audio))
        print_result(session.push_chunk(audio[start:end].tolist()), inline)

    final = session.finish()
    print_result(final, inline)
    return final


def run_microphone_streaming(pipe: Any, cfg: Any,
                            inline: bool, duration_sec: float, block_ms: int = 250) -> Any:
    try:
        import pyaudio
    except ModuleNotFoundError as exc:
        raise RuntimeError(
            "Microphone streaming requires the optional 'pyaudio' package. "
            "Install it with: pip install pyaudio"
        ) from exc

    session = pipe.create_streaming_session(cfg)

    block_size = max(1, int((block_ms / 1000.0) * SAMPLE_RATE))
    total_samples = int(duration_sec * SAMPLE_RATE)
    collected = 0

    p = pyaudio.PyAudio()
    stream = p.open(format=pyaudio.paInt16,
                   channels=1,
                   rate=SAMPLE_RATE,
                   input=True,
                   frames_per_buffer=block_size)

    print("Microphone is live. Start speaking now.")
    try:
        while collected < total_samples:
            remaining = total_samples - collected
            samples_to_read = min(block_size, remaining)
            data = stream.read(samples_to_read, exception_on_overflow=False)
            chunk = np.frombuffer(data, dtype=np.int16).astype(np.float32) / 32768.0
            print_result(session.push_chunk(chunk.tolist()), inline)
            collected += len(chunk)
    finally:
        stream.stop_stream()
        stream.close()
        p.terminate()

    final = session.finish()
    print_result(final, inline)
    return final


def build_streaming_config(args) -> Any:
    cfg = openvino_genai.ASRStreamingConfig()
    cfg.chunk_size_sec = args.chunk_size_sec
    cfg.warmup_chunks = args.warmup_chunks
    cfg.context_rollback_tokens = args.context_rollback_tokens
    return cfg


def main():
    parser = argparse.ArgumentParser(description="Streaming ASR sample for OpenVINO GenAI")
    parser.add_argument("model_dir", help="Path to the model directory")
    parser.add_argument("--device", default="CPU", help="Device to run the model on (default: CPU)")
    parser.add_argument("--chunk-size-sec", type=float, default=2.0,
                        help="Audio duration per decode pass (default: %(default)s)")
    parser.add_argument("--warmup-chunks", type=int, default=2,
                        help="Number of startup chunks without prefix (default: %(default)s)")
    parser.add_argument("--context-rollback-tokens", type=int, default=5,
                        help="How many trailing tokens to rewind when reusing prior output as prefix (default: %(default)s)")
    parser.add_argument("--wav", type=str, help="Path to a WAV file. If omitted, microphone input is used.")
    parser.add_argument("--duration", type=float,
                        help="Microphone capture duration in seconds. Required when --wav is not set. (default: %(default)s)")
    parser.add_argument("--block-ms", type=int, default=250,
                        help="Audio chunk size in milliseconds for both WAV and microphone streaming (default: %(default)s)")
    parser.add_argument("--inline", action="store_true",
                        help="Print only newly committed text without newlines between chunks")
    args = parser.parse_args()

    if args.wav is None and args.duration is None:
        parser.error("Either --wav or --duration must be provided. Use --duration for microphone mode.")

    cfg = build_streaming_config(args)
    pipe = openvino_genai.ASRPipeline(args.model_dir, args.device)

    if args.wav is not None:
        final = run_wav_streaming(pipe, args.wav, cfg, inline=args.inline, block_ms=args.block_ms)
    else:
        final = run_microphone_streaming(pipe, cfg, inline=args.inline, duration_sec=args.duration, block_ms=args.block_ms)

    if args.inline:
        print()  # newline after the inline committed text stream
    print(f"\n[final] ({final.language}) {final.committed_text}")


if __name__ == "__main__":
    main()
